includes/Core/DI/DI_Container.php

```diff
@@ -0,0 +1,196 @@
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\DI;
+
+use ArrayAccess;
+use Google\Site_Kit_Dependencies\Psr\Container\ContainerInterface;
+
+/**
+ * DI container class.
+ *
+ * @since n.e.x.t
+ */
+class DI_Container implements ContainerInterface, ArrayAccess {
+
+	/**
+	 * Definitions list.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	protected $definitions = array();
+
+	/**
+	 * Returns true if the container can return an entry for the given identifier.
+	 * Returns false otherwise.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to look for.
+	 * @return bool TRUE if the container can return an entry, otherwise FALSE.
+	 */
+	public function offsetExists( $id ) {
+		return $this->has( $id );
+	}
+
+	/**
+	 * Finds an entry of the container by its identifier and returns it.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to look for.
+	 * @return mixed Entry.
+	 */
+	public function offsetGet( $id ) {
+		return $this->get( $id );
+	}
+
+	/**
+	 * Sets the entry definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 * @param mixed  $entry Entry.
+	 */
+	public function offsetSet( $id, $entry ) {
+		$this->set( $id, $entry );
+	}
+
+	/**
+	 * Removes the entry.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to remove.
+	 */
+	public function offsetUnset( $id ) {
+		unset( $this->definitions[ $id ] );
+	}
+
+	/**
+	 * Sets the entry definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 * @param mixed  $entry Entry.
+	 * @return bool TRUE if the entry is added, otherwise FALSE.
+	 */
+	public function set( $id, $entry ) {
+		if ( ! empty( $this->definitions[ $id ]['is_protected'] ) ) {
+			return false;
+		}
+
+		$this->definitions[ $id ] = array(
+			'is_factory'   => false,
+			'is_protected' => false,
+			'is_service'   => is_callable( $entry ),
+			'entry'        => $entry,
+		);
+
+		return true;
+	}
+
+	/**
+	 * Sets the entry to be protected.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 */
+	public function set_protected( $id ) {
+		if ( $this->has( $id ) ) {
+			$this->definitions[ $id ]['is_protected'] = true;
+		}
+	}
+
+	/**
+	 * Sets the entry to be factory.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 */
+	public function set_factory( $id ) {
+		if ( $this->has( $id ) ) {
+			$this->definitions[ $id ]['is_factory'] = true;
+		}
+	}
+
+	/**
+	 * Sets services.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array $definitions Service definitions.
+	 */
+	public function set_services( array $definitions ) {
+		foreach ( $definitions as $service_name => $service_class ) {
+			$service_class_key = sprintf(
+				'%s_CLASS',
+				strtoupper( $service_name )
+			);
+
+			$this->set( $service_class_key, $service_class );
```

Comment:
Why is this needed?

Reply:
To let us easily override the class name that we want to use for a specific service. For example, let's assume we have these services:

```php
$di->set_services(
    array(
        'serviceA' => '\\Path\\To\\ClassA',
        'serviceB' => '\\Path\\To\\ClassB',
        ...
    )
)
```

Now, if we want to override `serviceA` and `serviceB`, and we don't use the "class name" approach, we need to do the following:

```php
$di['serviceA'] = function( $di ) {
    return new \Path\To\ClassA\SubclassA();
};

$di['serviceB'] = function( $di ) {
    return new \Path\To\ClassB\SubclassB();
};
```

To reduce the boilerplate, we can use the "class name" approach:

```php
$di['SERVICEA_CLASS'] = '\\Path\\To\\ClassA\\SubclassA';
$di['SERVICEB_CLASS'] = '\\Path\\To\\ClassB\\SubclassB';
```

Reply:
I think this is a bit overengineered. Overriding the callback function IMO isn't too much to ask, having to override the extra `%s_CLASS` key is actually less intuitive and more "magical" I think. Plus it wouldn't work anyway if a service needs constructor parameters.

IMO overriding the callback is more intuitive and more flexible, so I don't think we should store the class name in this extra key.

Reply:
Ok, updated it.